### ISSUE TITLE
Recommended practices for passing around Database objects and CouchResults

### ISSUE THREAD

**Issue by option-greek** [mibes/couch-rs#13]
Hi all,
When creating a db object through client.db() method, I'm overflowing stack quite frequently when calling from tokio workers. Is this expected (as in Database is a heavy object) and I should be using a Box ? I was theorizing that all couch-rs objects will be small in size as its just a http client to the couch.

Let me know what would be the recommended practice in this case.

**Comment by horacimacias** [comment]
I haven't seen this and I am using tokio as well. 

Do you have a sample code you could share that reproduces the problem?

**Comment by option-greek** [comment]
hi @horacimacias ,
    I'm posting the version of the code that works vs the one that doesn't (only difference is the Box). This code gets executed from warp web server which uses tokio for async. I don't get this error on the first call to get_db. It comes, when I call it a few times in the same tokio worker.


    pub async fn get_db(&self, db_key_orig: &String) -> Option<Box<Database>>
    {
        if db_key_orig.len() == 0 {
            println!("Trying to open a empty dbname {:?}", db_key_orig);
            return None;
        }

        let mut res = None;
        let db_key = self.form_db_name(db_key_orig);

        let cr = couch_rs::Client::new(&self.prefs.read().unwrap().db_host,
                                       &self.prefs.read().unwrap().db_user_name,
                                       &self.prefs.read().unwrap().db_pass);
        if cr.is_ok() {
            let client = cr.unwrap();

            let dbr
                = Box::new(client.db(&db_key).await);

            if dbr.is_ok() {
                let db = Box::new(dbr.unwrap());

                res = Option::from(db);
            } else {
                println!("Unable to create db {:?}. Err: {:?}", &db_key, dbr.err());
            }
        } else {
            println!("Unable to create cdb client. Err: {:?}", cr.err());
        }

        return res;
    }



    pub async fn get_db_no_box(&self, db_key_orig: &String) -> Option<Database>
    {
        if db_key_orig.len() == 0 {
            println!("Trying to open a empty dbname {:?}", db_key_orig);
            return None;
        }
        let mut res = None;
        let db_key = self.form_db_name(db_key_orig);

        let cr = couch_rs::Client::new(&self.prefs.read().unwrap().db_host,
                                       &self.prefs.read().unwrap().db_user_name,
                                       &self.prefs.read().unwrap().db_pass);
        if cr.is_ok() {
            let client = cr.unwrap();

            let dbr
                = client.db(&db_key).await;

            if dbr.is_ok() {
                let db = dbr.unwrap();

                res = Option::from(db);
            } else {
                println!("Unable to create db {:?}. Err: {:?}", &db_key, dbr.err());
            }
        } else {
            println!("Unable to create cdb client. Err: {:?}", cr.err());
        }

        return res;
    }

**Comment by horacimacias** [comment]
hi @option-greek ,

it's a bit hard to follow your example as I see it's relying on several other functions not provided. In any case, I really doubt the problem is related to Boxing or not Boxing the client.
If you want, you can run your program with `RUST_BACKTRACE=1` which will print a bit more information on the stack trace and you can copy paste the stack trace here.

I've taken your example and adapted it a bit to do something similar; perhaps you can use this to try at your end:

Cargo.toml
```toml
[package]
name = "testingcouchrs"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
couch_rs = "0.8.32"
tokio = { version = "1", features = ["full"] }
warp = "0.3.1"

```

main.rs
```rust
use std::convert::Infallible;

use warp::{Filter, Reply};

#[tokio::main]
async fn main() {
    // Match http://localhost:8080/open/dbname and open/create the database "dbname"
    let open_db = warp::get()
        .and(warp::path!("open" / String))
        .and_then(get_db);
    let routes = open_db;
    warp::serve(routes).run(([127, 0, 0, 1], 8080)).await;
}

pub async fn get_db(db_name: String) -> Result<impl Reply, Infallible> {
    if db_name.is_empty() {
        return Ok(warp::reply::html(
            "Trying to open a empty dbname".to_owned(),
        ));
    }

    match couch_rs::Client::new("http://localhost:5984", "admin", "password") {
        Ok(client) => match client.db(&db_name).await {
            Ok(db) => {
                return Ok(warp::reply::html(format!(
                    "Successfully created / opened db {}: {:?}",
                    &db_name, db
                )));
            }
            Err(e) => {
                return Ok(warp::reply::html(format!(
                    "Unable to create db {:?}. Err: {:?}",
                    &db_name, e
                )));
            }
        },
        Err(e) => {
            return Ok(warp::reply::html(format!(
                "Unable to connect. Err: {:?}",
                e
            )));
        }
    }
}
```

if you start couchdb on localhost:5984 with username "admin" and password "password", and run this rust program, you will be able to do a `curl http://localhost:8080/open/dbname` and this will create or open a database in couchdb.
I don't have any problem calling this multiple times.

As a side note, I see you're using `unwrap()` and `is_ok()`several times. I strongly suggest you use `match` instead or `if let Ok()` as it will make your code much more readable and idiomatic.
Still, this should make no difference to stacks overflowing but again I think it will help you focus on the rest of the code where the problem is likely to be.
Last but not least, I also suggest you run `cargo clippy` as many good suggestions will come from that.

**Comment by horacimacias** [comment]
☝️ you can start couchdb using

```bash
#!/bin/bash
set -e

COUCHDB_USER=admin
COUCHDB_PASSWORD=password

docker run --rm -p 5984:5984 -e COUCHDB_USER=$COUCHDB_USER -e COUCHDB_PASSWORD=$COUCHDB_PASSWORD couchdb:3
```

**Comment by mibes** [comment]
@option-greek: This example may also help you if you want to share the database across multiple REST requests:

```rust
use couch_rs::database::Database;
use serde_json::json;
use std::{convert::Infallible, sync::Arc};
use warp::{Filter, Reply};

const TEST_DB: &str = "test_db";

#[tokio::main]
async fn main() {
    start_warp().await;
}

pub async fn start_warp() {
    let client = couch_rs::Client::new_local_test().unwrap();
    let db = client.db(TEST_DB).await.unwrap();
    let db = Arc::new(db); // wrap it in an Arc to share

    let get_doc_path = warp::path!("api" / "v1" / "document" / String)
        .and(warp::get())
        .and(with_db(db))
        .and_then(|id: String, db: Arc<Database>| async move { get_doc(&id, db).await });

    let api = get_doc_path;
    warp::serve(api).bind(([0, 0, 0, 0], 8080)).await;
}

async fn get_doc(id: &str, db: Arc<Database>) -> Result<impl Reply, Infallible> {
    match db.get_raw(id).await {
        Ok(doc) => Ok(warp::reply::json(&doc)),
        Err(err) => Ok(warp::reply::json(&json!({"error": err.to_string()}))),
    }
}

fn with_db(
    db: Arc<Database>,
) -> impl Filter<Extract = (Arc<Database>,), Error = std::convert::Infallible> + Clone {
    warp::any().map(move || db.clone())
}
```

**Comment by option-greek** [comment]
Hi @horacimacias and @mibes , 
     The original issue occurs no more despite several retries on the test code. I'm assuming the original issue is probably that, in my code the function call to create the database is very deep down the chain of functions that result in the call (without all that nesting it works fine).

@mibes
I explored the approach of using Database with Arc earlier but the biggest issue is it works only in case the DB is already created before warp starts. Lets assume we have a dynamically created DB mapped to a key with the DB name. In that case, we will end up using a Hashmap to lookup the Database key and a RwLock to be able to insert the Database obect into the hashmap. Unfortunately it doesn't appear that the Database is 'Send' across the threads so when ever there is a await after the reference to the Arc<RwLock<HashMap<String, Database>>>, we get a compiler error stating that the RwLockGuard is not Send so we can't await. I slightly modified your code (which works) to demonstrate this case. Let me know if there is any way we can make the Database Send compliant.

```
use couch_rs::database::Database;
use serde_json::json;
use std::{convert::Infallible, sync::Arc};
use warp::{Filter, Reply};
use std::sync::RwLock;
use std::collections::HashMap;

const TEST_DB: &str = "test_db";

#[tokio::main]
async fn main() {
    start_warp().await;
}

pub async fn start_warp() {
    let db_list = Arc::new(RwLock::new(HashMap::new()));

    let get_doc_path = warp::path!("api" / "v1" / "document" / String)
        .and(warp::get())
        .and(with_db(db_list))
        .and_then(|id: String, db: Arc<RwLock<HashMap<String, Database>>>| async move { get_doc(&id, db).await });

    let api = get_doc_path;
    warp::serve(api).bind(([0, 0, 0, 0], 8080)).await;
}

async fn get_doc(id: &str, mut db_list: Arc<RwLock<HashMap<String, Database>>>) -> Result<impl Reply, Infallible> {
    let mut db_list_ref = db_list.write().unwrap();
    if !db_list_ref.contains_key(TEST_DB) {
        let client = couch_rs::Client::new_local_test().unwrap();
        let db = client.db(TEST_DB).await.unwrap();
        db_list_ref.insert(TEST_DB.to_string(), db );

    }

    let db = db_list_ref.get(&TEST_DB.to_string()).unwrap();
    match db.get_raw(id).await {
        Ok(doc) => Ok(warp::reply::json(&doc)),
        Err(err) => Ok(warp::reply::json(&json!({"error": err.to_string()}))),
    }
}

fn with_db(
    db: Arc<RwLock<HashMap<String, Database>>>,
) -> impl Filter<Extract = (Arc<RwLock<HashMap<String, Database>>>,), Error = std::convert::Infallible> + Clone {
    warp::any().map(move || db.clone())
}

```

Error is:

error: future cannot be sent between threads safely
  --> src\main.rs:21:10
   |
21 |         .and_then(|id: String, db: Arc<RwLock<HashMap<String, Database>>>| async move { get_doc(&id, db).await });
   |          ^^^^^^^^ future created by async block is not `Send`
   |
   = help: within `impl Future`, the trait `Send` is not implemented for `std::sync::RwLockWriteGuard<'_, HashMap<std::string::String, Database>>`
note: future is not `Send` as this value is used across an await
  --> src\main.rs:37:11
   |
28 |     let mut db_list_ref = db_list.write().unwrap();
   |         --------------- has type `std::sync::RwLockWriteGuard<'_, HashMap<std::string::String, Database>>` which is not `Send`
...
37 |     match db.get_raw(id).await {
   |           ^^^^^^^^^^^^^^^^^^^^ await occurs here, with `mut db_list_ref` maybe used later
...
41 | }
   | - `mut db_list_ref` is later dropped here

error: aborting due to previous error; 1 warning emitted

**Comment by horacimacias** [comment]
I think your issue is you're using `std::sync::RwLock` which is not Send.
The Database itself is Send.

If you use `tokio::sync::RwLock` instead (you'll need `db_list.write().await` instead of `db_list.write().unwrap()`) the code should compile.

**Comment by option-greek** [comment]
So that was the issue then. It works perfect now! (I'm new to rust and had no clue that both RwLocks are different)

Thank you @horacimacias and @mibes for your patience and help.

**Comment by horacimacias** [comment]
happy to help!

I suggest you also have a look at the tokio tutorial; I found it quite useful. E.g. https://tokio.rs/tokio/tutorial/shared-state
And remember, `clippy` is your friend!
